### DISCUSSION TITLE
Adjoint of transformix transformation

### DISCUSSION THREAD

**Discussion by gschramm** [SuperElastix/elastix#883] [Q&A]
Hi all,

I am new to elastix and transformix and was wondering if it is possible to get the exact adjoint (not inverse)
of an image transformation used by transformix.

I am working on an iterative optimization algorithm where the adjoint of the transformation operator is needed.
The adjoint is defined via <Tx,y> = <x,T^H y> where T^H is the adjoint of T.

Georg

**Comment by stefanklein** [comment]
HI Georg, 
Welcome to Elastix! There is no direct way to get the adjoint, but you can extract the transformation matrix and compute the adjoint yourself. 
Best! Stefan

**Reply by gschramm** [reply]
Hi Stefan,

many thanks for the feedback! Is there a short tutorial that shows how to extract the transformation matrix?

Georg

**Reply by ntatsisk** [reply]
Hi @gschramm,

If you happen to use python then the following (still draft PR) notebook might be interesting to you: https://github.com/InsightSoftwareConsortium/ITKElastix/pull/193

The idea is summarized in these lines of the notebook:
```
# Register 
result_image, result_transform_parameters = itk.elastix_registration_method(
    fixed_image, moving_image, parameter_object=parameter_object)

# Extract useful transformation parameters
parameter_map = result_transform_parameters.GetParameterMap(0)
center_of_rotation = np.array(parameter_map['CenterOfRotationPoint'], dtype=float).tolist()

# Convert paramters from Elastix format to ITK 3x3 matrix and 3-element translation vector
transform_parameters = np.array(parameter_map['TransformParameters'], dtype=float)
matrix = transform_parameters[:ndim*ndim].reshape(ndim, ndim)
translation = transform_parameters[-ndim:].tolist()
```
where `ndim` is the dimensionality of the images (2 for 2D, 3 for 3D). For example for a 3D, `matrix` will be a 3x3 affine matrix (scaling, shearing, rotations) and `translation` a 3-element translation vector. If you would like to have both them stored in a single 4x4 matrix so that you invert it then you can do:

```
affine_matrix = np.zeros((ndim+1, ndim+1), dtype=np.float32)
affine_matrix[:ndim, :ndim] = matrix
affine_matrix[:ndim, ndim] = translation
```
It assumes that you are using an affine transformation though. Hope this helps!

**Reply by gschramm** [reply]
Thanks a lot! Unfortunately, I am using a non-affine transformation.

**Reply by ntatsisk** [reply]
If by "non-affine" you still mean any of the Translation/Rigid/Euler/Similarity transforms, I guess the code above can be adjusted to work for those cases. However, if you mean b-spline transform (which is non-linear) then I would assume that the adjoint matrix is not defined for that case. Happy to be corrected though!

**Reply by gschramm** [reply]
I guess that depends on the degree of the B-spline. A degree 2 B-spline should use linear functions, so I guess if I can choose the degree of the spline, it should be possible that the transformation is a linear map.
